File: src/main/java/io/protobase/f7/formulas/parser/TO_TEXT.java
package io.protobase.f7.formulas.parser;

import io.protobase.f7.errors.F7Exception;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class TO_TEXT extends AbstractFormula {
  public static FormulaName NAME = FormulaName.TO_PERCENT;
  public static TO_TEXT SELF = new TO_TEXT();

  public TO_TEXT() {
    super();
  }

  public TO_TEXT(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    Object value = Converters.first(collateralLookup.apply(origin, values[0]));
    if (value instanceof F7Exception) {
      return value;
    }
    return Converters.toText(value);
  }
}

File: src/main/js/formulas/parser/TO_TEXT.ts
import { F7Exception } from "../../errors/F7Exception";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class TO_TEXT extends AbstractFormula {
  static SELF: TO_TEXT = new TO_TEXT();
  NAME = FormulaName.TO_TEXT;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    const first = Converters.first(this.collateralLookup(origin, values[0]));
    if (first instanceof F7Exception) {
      return first;
    }
    return Converters.toText(first);
  }
}
